File: IdentityX/testing_framework/sessions_test.go
package testing

import (
	"GoAuth/internal/errx"
	"context"
	"net/http"
	"testing"
)

func testSessions(t *testing.T, suite *TestSuite) {
	// Create user in parent test context
	client := suite.NewClient(t)
	user := client.WithCredentials("sessions@mail.com", ValidPassword).Register().Login()

	t.Run("ListSessions", func(t *testing.T) {
		// Create a new client with subtest's t for the authenticated client
		// We need to preserve the auth context but use the new t
		authClient := suite.NewClient(t).WithAuth(user.auth)
		authClient.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(1)
	})

	t.Run("MultipleLoginsSessions", func(t *testing.T) {
		// Create 3 more sessions (we already have 1)
		client2 := suite.NewClient(t)
		client2.WithCredentials(user.email, user.password).Login()

		client3 := suite.NewClient(t)
		client3.WithCredentials(user.email, user.password).Login()

		client4 := suite.NewClient(t)
		user4 := client4.WithCredentials(user.email, user.password).Login()

		arr := user4.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray()

		arr.Length().IsEqual(4)

		// Get oldest session ID to revoke
		currentSessionID := arr.Value(0).Object().Value("session_id").String().Raw()
		oldestSessionID := arr.Value(3).Object().Value("session_id").String().Raw()

		// Can't revoke current session
		user4.DELETE("/sessions/" + currentSessionID).
			Expect(http.StatusForbidden).
			HasErrID(errx.SessionSelfRevokeForbidden).
			HasMessage("cannot revoke the currently active session")

		// Revoke first session
		user4.DELETE("/sessions/" + oldestSessionID).
			Expect(http.StatusOK).
			HasMessage("revoked session")

		// Should have 3 sessions now
		user4.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(3)
	})

	t.Run("RevokeOtherSessions", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("revoke-others@mail.com", ValidPassword).Register().Login()

		// Create more sessions
		suite.NewClient(t).WithCredentials(user.email, user.password).Login()
		suite.NewClient(t).WithCredentials(user.email, user.password).Login()

		user.DELETE("/sessions/others").
			Expect(http.StatusOK).
			HasMessage("revoked sessions")

		// Should only have current session
		user.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(1)
	})

	t.Run("SessionInfo", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("session-me@mail.com", ValidPassword).Register().Login()

		data := user.GET("/sessions/me").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"refresh_expire_date": AnyNumber{},
			"access": map[string]interface{}{
				"iss": "GoAuth",
				"exp": AnyNumber{},
				"iat": AnyNumber{},
				"jti": AnyUUID{},
				"sub": map[string]interface{}{
					"id":         AnyUUID{},
					"email":      "session-me@mail.com",
					"project_id": nil,
					"user_type":  "client",
					"metadata":   nil,
					"session_id": AnyUUID{},
					"user_agent": AnyString{},
					"user_ip":    AnyString{},
				},
			},
		}

		Validate(t, data, spec)
	})

	t.Run("RevokeAllSessions", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("revoke-all@mail.com", ValidPassword).Register().Login()

		// Create more sessions
		suite.NewClient(t).WithCredentials(user.email, user.password).Login()
		suite.NewClient(t).WithCredentials(user.email, user.password).Login()

		user.DELETE("/sessions").
			Expect(http.StatusOK).
			HasMessage("revoked sessions")

		// Session should be invalid
		user.GET("/sessions").
			Expect(http.StatusUnauthorized).
			HasErrID(errx.SessionRevoked).
			HasMessage("session not found or revoked")
	})

	// ExpiredSessionNotListed tests that expired sessions are not returned when listing active sessions.
	// This ensures that only currently valid sessions are visible to the user.
	// The test manually inserts an expired session directly into the database for a registered user.
	// It then attempts to list sessions for that user, expecting only the active session (created by login)
	// to be returned, and not the manually inserted expired one.
	t.Run("ExpiredSessionNotListed", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("expired@mail.com", ValidPassword).Register().Login()

		ctx := context.Background()

		// Manually insert an expired session for this user
		var userID string
		err := suite.DB.QueryRow(ctx, "SELECT id FROM users WHERE email = 'expired@mail.com'").Scan(&userID)
		if err != nil {
			t.Fatalf("Failed to get user ID: %v", err)
		}

		var identityID string
		err = suite.DB.QueryRow(ctx, `
			INSERT INTO identities (type, entity_id)
			VALUES ('client', $1)
			ON CONFLICT (type, entity_id) DO UPDATE SET type = 'client'
			RETURNING id
		`, userID).Scan(&identityID)
		if err != nil {
			t.Fatalf("Failed to create session identity: %v", err)
		}

		_, err = suite.DB.Exec(ctx, `
			INSERT INTO sessions (
				identity_id, issued_at, user_agent, user_ip, expires_at, created_at, updated_at, user_type
			) VALUES (
				$1,
				NOW() - INTERVAL '2 days',
				'Expired Agent',
				'127.0.0.1',
				NOW() - INTERVAL '1 day',
				NOW(),
				NOW(),
				'client'
			)
		`, identityID)
		if err != nil {
			t.Fatalf("Failed to insert expired session: %v", err)
		}

		// Verify that the expired session is NOT in the list
		// Should only have the active login session (1), ignoring the manually inserted expired one
		user.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(1)
	})

	// RevokedSessionNotListed tests that manually revoked sessions are not returned when listing active sessions.
	// This verifies that sessions explicitly marked as revoked are correctly filtered out.
	// The test manually inserts a revoked session directly into the database for a registered user.
	// It then lists sessions for that user, expecting only the active session (created by login)
	// to be returned, and not the manually inserted revoked one.
	t.Run("RevokedSessionNotListed", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("revoked@mail.com", ValidPassword).Register().Login()

		ctx := context.Background()

		// Manually insert a revoked session for this user
		var userID string
		err := suite.DB.QueryRow(ctx, "SELECT id FROM users WHERE email = 'revoked@mail.com'").Scan(&userID)
		if err != nil {
			t.Fatalf("Failed to get user ID: %v", err)
		}

		var identityID string
		err = suite.DB.QueryRow(ctx, `
			INSERT INTO identities (type, entity_id)
			VALUES ('client', $1)
			ON CONFLICT (type, entity_id) DO UPDATE SET type = 'client'
			RETURNING id
		`, userID).Scan(&identityID)
		if err != nil {
			t.Fatalf("Failed to create session identity: %v", err)
		}

		_, err = suite.DB.Exec(ctx, `
			INSERT INTO sessions (
				identity_id, issued_at, user_agent, user_ip, revoked_at, created_at, updated_at, expires_at, user_type
			) VALUES (
				$1,
				NOW() - INTERVAL '2 days',
				'Expired Agent',
				'127.0.0.1',
				NOW() - INTERVAL '1 day',
				NOW(),
				NOW(),
				NOW() + INTERVAL '1 day',
				'client'
			)
		`, identityID)
		if err != nil {
			t.Fatalf("Failed to insert revoked session: %v", err)
		}

		// Verify that the revoked session is NOT in the list
		// Should only have the active login session (1), ignoring the manually inserted revoked one
		user.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(1)
	})

	// SessionLeakage tests that session isolation is correctly enforced between different identity types (client vs. project user).
	// This is a critical security test to prevent one user type from accessing or even seeing sessions belonging to another type,
	// even if both identity types might logically be linked to the same underlying individual or email.
	// The test creates a client user and a project user (both potentially using the same email).
	// It then logs in as each user, creating separate sessions for each identity type.
	// Finally, it attempts to list sessions while authenticated as the client user,
	// expecting to see only the client user's session and ensuring that the project user's session is not exposed.
	t.Run("SessionLeakage", func(t *testing.T) {
		// This test ensures that a user with a 'client' identity cannot see sessions
		// from a 'project' identity, even if they are logically the same user.

		// 1. Create a client user
		clientUser := suite.NewClient(t).WithCredentials("leakage-client@mail.com", ValidPassword)
		clientUser.Register()

		// 2. Create a project and a project user with the same email
		projectOwner := suite.NewClient(t).WithCredentials("leakage-project-owner@mail.com", ValidPassword)
		project := projectOwner.Register().Login().CreateProject("Leakage Test Project")

		projectUser := suite.NewClient(t).WithCredentials("leakage-client@mail.com", ValidPassword)
		projectUser.ProjectRegister(project.projectID)

		// 3. Log in as both the client and project user to create two separate sessions
		clientSession := clientUser.Login()
		projectUser.ProjectLogin(project.projectID) // This creates a session for the project user

		// 4. As the client user, list the sessions
		// We expect to only see the client user's session, not the project user's.
		sessions := clientSession.GET("/sessions").
			Expect(http.StatusOK).
			RequireDataArray()

		// 5. Verify that only one session is returned
		sessions.Length().IsEqual(1)

		// Optional: Verify the user_type of the returned session
		sessionType := sessions.Value(0).Object().Value("user_type").String().Raw()
		if sessionType != "client" {
			t.Errorf("Expected user_type to be 'client', but got '%s'", sessionType)
		}
	})
}

File: IdentityX/testing_framework/sessions.test.ts
import { afterAll, beforeAll, describe, expect, test } from "vitest";
import { createClient } from "./helpers/index.js";
import { assertErrID, assertMessage, shouldFail } from "./helpers/assert.js";
import { Validate, AnyUUID, AnyString, AnyNumber } from "./helpers/validate.js";
import { dbQuery, closeDB } from "./helpers/db.js";
import { ValidPassword } from "./fixtures/auth/testdata.js";

// Error IDs
// SessionSelfRevokeForbidden = fail.ID(0, "SESSION", 1, true, ...) → 0_SESSION_0001_S
// SessionRevoked             = fail.ID(0, "SESSION", 0, false, ...) → 0_SESSION_0000_D
const ErrSessionSelfRevokeForbidden = "0_SESSION_0001_S";
const ErrSessionRevoked             = "0_SESSION_0000_D";

afterAll(async () => {
    await closeDB();
});

// ============================================================================
// SESSIONS TESTS
// ============================================================================

describe("Sessions", () => {
    let sharedUser: Awaited<ReturnType<typeof createClient>>;

    beforeAll(async () => {
        await createClient().withCredentials("sessions@mail.com", ValidPassword).register();
        sharedUser = await createClient().withCredentials("sessions@mail.com", ValidPassword).login();
    });

    test("ListSessions", async () => {
        const sessions = await sharedUser.get("/sessions");
        expect(sessions).toHaveLength(1);
    });

    test("MultipleLoginsSessions", async () => {
        await createClient().withCredentials("sessions@mail.com", ValidPassword).login();
        await createClient().withCredentials("sessions@mail.com", ValidPassword).login();
        const user4 = await createClient().withCredentials("sessions@mail.com", ValidPassword).login();

        const sessions: any[] = await user4.get("/sessions");
        expect(sessions).toHaveLength(4);

        const currentSessionID = sessions[0].session_id;
        const oldestSessionID  = sessions[3].session_id;

        // Can't revoke current session
        const forbiddenBody = await shouldFail(user4.del(`/sessions/${currentSessionID}`), 403);
        assertErrID(forbiddenBody, ErrSessionSelfRevokeForbidden);
        assertMessage(forbiddenBody, "cannot revoke the currently active session");

        // Revoke oldest session
        await user4.del(`/sessions/${oldestSessionID}`);

        // Should have 3 sessions now
        const updated: any[] = await user4.get("/sessions");
        expect(updated).toHaveLength(3);
    });

    test("RevokeOtherSessions", async () => {
        await createClient().withCredentials("revoke-others@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("revoke-others@mail.com", ValidPassword).login();

        await createClient().withCredentials("revoke-others@mail.com", ValidPassword).login();
        await createClient().withCredentials("revoke-others@mail.com", ValidPassword).login();

        await user.del("/sessions/others");

        const sessions: any[] = await user.get("/sessions");
        expect(sessions).toHaveLength(1);
    });

    test("SessionInfo", async () => {
        await createClient().withCredentials("session-me@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("session-me@mail.com", ValidPassword).login();

        const data = await user.get("/sessions/me");

        Validate(data, {
            refresh_expire_date: AnyNumber,
            access: {
                iss: "GoAuth",
                exp: AnyNumber,
                iat: AnyNumber,
                jti: AnyUUID,
                sub: {
                    id:         AnyUUID,
                    email:      "session-me@mail.com",
                    project_id: null,
                    user_type:  "client",
                    metadata:   null,
                    session_id: AnyUUID,
                    user_agent: AnyString,
                    user_ip:    AnyString,
                },
            },
        });
    });

    test("RevokeAllSessions", async () => {
        await createClient().withCredentials("revoke-all@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("revoke-all@mail.com", ValidPassword).login();

        await createClient().withCredentials("revoke-all@mail.com", ValidPassword).login();
        await createClient().withCredentials("revoke-all@mail.com", ValidPassword).login();

        await user.del("/sessions");

        const body = await shouldFail(user.get("/sessions"), 401);
        assertErrID(body, ErrSessionRevoked);
        assertMessage(body, "session not found or revoked");
    });

    test("ExpiredSessionNotListed", async () => {
        await createClient().withCredentials("expired@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("expired@mail.com", ValidPassword).login();

        const { rows: [{ id: userID }] } = await dbQuery<{ id: string }>(
            "SELECT id FROM users WHERE email = 'expired@mail.com'"
        );

        const { rows: [{ id: identityID }] } = await dbQuery<{ id: string }>(
            `INSERT INTO identities (type, entity_id)
       VALUES ('client', $1)
       ON CONFLICT (type, entity_id) DO UPDATE SET type = 'client'
       RETURNING id`,
            [userID]
        );

        await dbQuery(
            `INSERT INTO sessions (identity_id, issued_at, user_agent, user_ip, expires_at, created_at, updated_at, user_type)
       VALUES ($1, NOW() - INTERVAL '2 days', 'Expired Agent', '127.0.0.1', NOW() - INTERVAL '1 day', NOW(), NOW(), 'client')`,
            [identityID]
        );

        const sessions: any[] = await user.get("/sessions");
        expect(sessions).toHaveLength(1);
    });

    test("RevokedSessionNotListed", async () => {
        await createClient().withCredentials("revoked@mail.com", ValidPassword).register();
        const user = await createClient().withCredentials("revoked@mail.com", ValidPassword).login();

        const { rows: [{ id: userID }] } = await dbQuery<{ id: string }>(
            "SELECT id FROM users WHERE email = 'revoked@mail.com'"
        );

        const { rows: [{ id: identityID }] } = await dbQuery<{ id: string }>(
            `INSERT INTO identities (type, entity_id)
       VALUES ('client', $1)
       ON CONFLICT (type, entity_id) DO UPDATE SET type = 'client'
       RETURNING id`,
            [userID]
        );

        await dbQuery(
            `INSERT INTO sessions (identity_id, issued_at, user_agent, user_ip, revoked_at, created_at, updated_at, expires_at, user_type)
       VALUES ($1, NOW() - INTERVAL '2 days', 'Revoked Agent', '127.0.0.1', NOW() - INTERVAL '1 day', NOW(), NOW(), NOW() + INTERVAL '1 day', 'client')`,
            [identityID]
        );

        const sessions: any[] = await user.get("/sessions");
        expect(sessions).toHaveLength(1);
    });

    test("SessionLeakage", async () => {
        await createClient().withCredentials("leakage-client@mail.com", ValidPassword).register();
        await createClient().withCredentials("leakage-project-owner@mail.com", ValidPassword).register();

        const owner = await createClient()
            .withCredentials("leakage-project-owner@mail.com", ValidPassword)
            .login();

        const project = await owner.post("/projects", {
            project_name: "Leakage Test Project",
            metadata: { env: "test" },
        });
        const projectID = project.id;

        // Register and log in client user as project user
        await createClient().http.post(`/projects/${projectID}/register`, {
            email: "leakage-client@mail.com",
            password: ValidPassword,
        });

        const clientSession = await createClient()
            .withCredentials("leakage-client@mail.com", ValidPassword)
            .login();

        // Create a project user session (separate identity type)
        await createClient().http.post(`/projects/${projectID}/login`, {
            email: "leakage-client@mail.com",
            password: ValidPassword,
        });

        // Client should only see their own session, not the project user's
        const sessions: any[] = await clientSession.get("/sessions");
        expect(sessions).toHaveLength(1);
        expect(sessions[0].user_type).toBe("client");
    });
});